### PR TITLE
rgw: fixes for virtual hosting of buckets

### PR DESCRIPTION
**Pull request by cbodley**
this brings back the fix from @robbat2 that was reverted in https://github.com/ceph/ceph/pull/10873, with an additional check that disables the virtual hosting feature if no hostnames were configured. this prevents default configurations (such as teuthology) from treating all requests as virtual hosted buckets

also updated the docs - a preview should eventually be available at http://docs.ceph.com/docs/wip-17440/radosgw/s3/commons/

### CODE REVIEW

File: src/rgw/rgw_rest.cc [hunk 1]
```diff
@@ -1846,10 +1846,12 @@ int RGWREST::preprocess(struct req_state *s, RGWClientIO* cio)
     // As additional checks:
     // - if the Host header is an IP, we're using path-style access without DNS
     // - Also check that the Host header is a valid bucket name before using it.
+    // - Don't enable virtual hosting if no hostnames are configured
     if (subdomain.empty()
         && (domain.empty() || domain != info.host)
         && !looks_like_ip_address(info.host.c_str())
-        && RGWHandler_REST::validate_bucket_name(info.host)) {
+        && RGWHandler_REST::validate_bucket_name(info.host) == 0
+        && !hostnames_set.empty()) {
```
Comment (robbat2) [inline src/rgw/rgw_rest.cc]: This should check hostnames_s3website_set as well. If either of them are non-empty, it should be ok (should be very rare, but want to check anyway).

Comment (cbodley) [inline src/rgw/rgw_rest.cc]: added

**Comment by robbat2** [comment]
I ran into one related impact from this, that should be clarified & documented.
RGW's legacy behavior differs from S3 behavior, and s3-tests was designed around the RGW legacy behavior: using the path calling format, and assuming that the hostname it was connecting to was also configured as the rgw dns name or zonegroup hostname.

Thus, in some cases, we know get a lot more false-positive s3-tests breakage.

```
GET /foobar/ HTTP/1.1
Host: non-existing-ceph-buckettest.example.com
```

AWS-S3 treats this as bucket {{non-existing-ceph-buckettest.example.com}}, and returns:
```
<?xml version="1.0" encoding="UTF-8"?>
<Error>
<Code>NoSuchBucket</Code>
<Message>The specified bucket does not exist</Message>
<BucketName>non-existing-ceph-buckettest.example.com</BucketName>
<RequestId>D1F9F7DCD9D79A11</RequestId>
<HostId>1RGXmTwGt4w2R+BV/xiuJsmJoIwVZsjhwEBH2Qrqh4URbV0zlIvX1+9KczHCCKHnhtV7Mhh7Ps8=</HostId>
</Error>
```

RGW, prior to my patching, treats this as bucket {{foobar}}, and returns:
```
<?xml version="1.0" encoding="UTF-8"?>
<Error>
<Code>NoSuchBucket</Code>
<BucketName>foobar</BucketName>
<RequestId>tx0000000000000000025fb-0057f2e73e-fa74-default</RequestId>
<HostId>fa74-default-default</HostId>
</Error>
```

RGW, with patch, now matches AWS-S3:
```
<?xml version="1.0" encoding="UTF-8"?>
<Error>
<Code>NoSuchBucket</Code>
<BucketName>non-existing-ceph-buckettest.example.com</BucketName>
<RequestId>tx0000000000000000c6085-0057f2e721-10bb00c9-default</RequestId>
<HostId>10bb00c9-default-default</HostId>
</Error>
```

**Comment by cbodley** [comment]
oops, i was cleaning up branches and nuked this one. reopened

**Comment by robbat2** [comment]
@cbodley @yehudasa what happened to having this merged in v10.2.4? I'm writing documentation for StaticSites, and it depends on having this merged.

**Comment by oritwas** [comment]
I scheduled a teuthology run: http://pulpito.ceph.com/owasserm-2016-10-27_16:41:06-rgw-wip-17440---basic-mira/

**Comment by oritwas** [comment]
jenkins test this please

**Comment by oritwas** [comment]
new run (previous runs had enviorment problems): http://pulpito.ceph.com/owasserm-2016-10-31_18:27:38-rgw-wip-orit-17440---basic-mira/

**Comment by cbodley** [comment]
thanks @oritwas !

**Pull request merged**